### PR TITLE
NATIVE-164: Persist city selection

### PR DESCRIPTION
**Pull request by klinzo**

### CODE REVIEW

File: src/routes/landing/components/Landing.js [hunk 1]
```diff
@@ -25,12 +26,35 @@ export type PropsType = {
...
+    const { language, navigateToDashboard } = this.props
+    const localizationSettings = new LocalizationSettings()
+
+    const selectedCity = await localizationSettings.loadSelectedCity()
+
+    if (selectedCity) {
+      navigateToDashboard(selectedCity, language)
+    } else {
+      this.setState({localizationLoaded: true})
+    }
```
Comment (Schedulaar) [inline src/routes/landing/components/Landing.js]: I think this switch should be already done in the Navigator to avoid jumpy screens at the start of the app and not even navigate to the LandingPage first

Comment (klinzo) [reply src/routes/landing/components/Landing.js]: Done. It adds a bit more complexity and there is still kind of a jumpy screen, but should be ok.

File: src/modules/app/components/Navigator.js [hunk 2]
```diff
@@ -74,12 +78,85 @@ export const AppStack = createStackNavigator(
...
+      throw new Error('ContentLanguage must be set!')
+    } else if (!this.navigator) {
+      throw new Error('Ref must not be null!')
+    }
+
+    if (selectedCity) {
+      this.navigateToDashboard(selectedCity, contentLanguage)
+    } else {
+      // $FlowFixMe dispatch is missing in type
+      this.navigator.dispatch(NavigationActions.navigate({
```
Comment (Schedulaar) [inline src/modules/app/components/Navigator.js]: Can't you do sth like `this.navigator.navigate('Landing')`?

File: src/routes/landing/containers/LandingContainer.js [hunk 3]
```diff
@@ -48,28 +47,23 @@ const mapStateToProps = (state: StateType): StatePropsType => {
...
   return {
-    fetchCities: () => dispatch({
-      type: 'FETCH_CITIES'
-    }),
     navigateToDashboard: (cityCode: string, language: string) => {
       const path = `/${cityCode}/${language}`
       const key: string = generateKey()
 
-      const action: NavigationReplaceAction = StackActions.replace({
+      const action = NavigationActions.navigate({
```
Comment (Schedulaar) [inline src/routes/landing/containers/LandingContainer.js]: Revert this change; We need to set newKey here and reset the stack:

We need react-navigation to give this route the very same key that we generate above. (Otherwise there's some strange default key, which messes things up)

File: src/modules/app/components/Navigator.js [hunk 4]
```diff
@@ -74,12 +78,85 @@ export const AppStack = createStackNavigator(
...
+      params: {
+        cityCode,
+        key,
+        sharePath: path,
+        onRouteClose: () => this.props.clearCategory(key)
+      }
+    })
+
+    // $FlowFixMe dispatch is missing in type, https://github.com/react-navigation/react-navigation/issues/3842
+    this.navigator.dispatch(NavigationActions.navigate({
```
Comment (Schedulaar) [inline src/modules/app/components/Navigator.js]: I think we should not avoid a direct `dispatch` where possible...
```js
this.navigator.navigate({
  routeName: 'App',
  action: navigateToDashboard
})
```
should do the trick

File: src/modules/app/components/Navigator.js [hunk 5]
```diff
@@ -74,12 +78,85 @@ export const AppStack = createStackNavigator(
...
+        routeName: 'Landing'
+      }))
+    }
+  }
+
+  navigateToDashboard (cityCode: string, language: string) {
+    const path = `/${cityCode}/${language}`
+    const key = generateKey()
+
+    const navigateToDashboard = NavigationActions.navigate({
```
Comment (Schedulaar) [inline src/modules/app/components/Navigator.js]: I guess we need to do that with StackReplace action (as it was done in the Landing Page)

File: src/modules/app/components/Navigator.js (hunk 2 above)
Comment (klinzo) [reply src/modules/app/components/Navigator.js]: See https://reactnavigation.org/docs/en/app-containers.html#docsNav . `navigator.navigate()` is not working, I've tried it.

File: src/modules/app/components/Navigator.js [hunk 6]
```diff
@@ -74,12 +78,85 @@ export const AppStack = createStackNavigator(
...
+  localizationSettings: LocalizationSettings
+  navigator: ?NavigationContainer<NavigationState, {}, {}>
+
+  constructor (props: PropsType) {
+    super(props)
+    this.localizationSettings = new LocalizationSettings()
+  }
+
+  componentWillMount () {
+    this.props.fetchCities()
```
Comment (Schedulaar) [inline src/modules/app/components/Navigator.js]: Why do we need to fetchCities here?

File: src/modules/app/components/Navigator.js [hunk 7]
```diff
@@ -74,12 +78,85 @@ export const AppStack = createStackNavigator(
...
+    this.navigator.dispatch(NavigationActions.navigate({
+      routeName: 'App',
+      action: navigateToDashboard
+    }))
+
+    this.props.fetchCategory(cityCode, language, key)
+  }
+
+  render () {
+    // eslint-disable-next-line react/jsx-no-bind
```
Comment (Schedulaar) [inline src/modules/app/components/Navigator.js]: I think we should not disable this error here: Every time render is called, the ref prop will always get updated, because the arrow function is generated in every `render` iteration

File: src/routes/landing/containers/LandingContainer.js (hunk 3 above)
Comment (klinzo) [reply src/routes/landing/containers/LandingContainer.js]: Very strange, but okay :D Thanks :)

File: src/modules/app/components/Navigator.js (hunk 4 above)
Comment (klinzo) [reply src/modules/app/components/Navigator.js]: See above.

File: src/modules/app/components/Navigator.js (hunk 6 above)
Comment (klinzo) [reply src/modules/app/components/Navigator.js]: Because the fetch is otherwise never called, this was done in the Landing page before but since we skip the Landing page if there is a selected city, the fetch will not happen. We could of course also do this in `loadCityContent`

Comment (klinzo) [reply src/modules/app/components/Navigator.js]: But then we would have to call it at two places.

**Comment by klinzo** [comment]
@Schedulaar how would injecting an initial route config solve the problem that we don't need a Loading screen? The loading route is for the time it takes to decide whether there is a selected city or not. If we set an initial route config, be have the problem with jumpy screens again. Am I understanding something wrong here?

File: src/modules/app/StateType.js [hunk 8]
```diff
@@ -10,7 +10,7 @@ import {
   LanguageModel
 } from '@integreat-app/integreat-api-client'
 import Moment from 'moment'
-import { DEFAULT_LANGUAGE } from '../i18n/components/I18nProvider'
```
Comment (maxammann) [inline src/modules/app/StateType.js]: I would prefere to stick to the i18n name. It is the default for basically every software project.

Also we should try not to rename files if not really needed. (Because of merge conflicts etc.)

File: src/modules/app/components/Navigator.js [hunk 9]
```diff
@@ -74,12 +79,86 @@ export const AppStack = createStackNavigator(
...
+|}
+
+class Navigator extends React.Component<PropsType> {
+  localizationSettings: LocalizationSettings
+  navigator: {current: null | React$ElementRef<NavigationContainer<NavigationState, {}, {}>>}
+
+  constructor (props: PropsType) {
+    super(props)
+    this.localizationSettings = new LocalizationSettings()
+    this.navigator = React.createRef()
```
Comment (maxammann) [inline src/modules/app/components/Navigator.js]: I'm pretty sure we can move the `navigator` reference to a child of this component.

Can the function `loadSelectedCity` be called in the `componentWillMount` of a child component? Then you can use the `withNavigation` hoc to get a reference to the navigator

File: src/modules/localization/LocalizationSettings.js [hunk 10]
```diff
@@ -18,6 +19,18 @@ class LanguageAsyncStorage {
...
+    await this.asyncStorage.setItem(SELECTED_CITY, city)
+  }
+
+  clearSelectedCity = async () => {
+    await this.asyncStorage.removeItem(SELECTED_CITY)
+  }
 }
 
-export default LanguageAsyncStorage
+export default LocalizationSettings
```
Comment (maxammann) [inline src/modules/localization/LocalizationSettings.js]: Should be moved to AppSettings as discussed in call.

File: src/modules/app/components/Navigator.js (hunk 9 above)
Comment (klinzo) [reply src/modules/app/components/Navigator.js]: Could you elaborate further what you mean by that? If I understand you correctly, you propose to introduce a child component to handle the navigating as soon as it is known whether a city is selected or not.
Afaics you can't do any navigating outside the `AppContainer` without using a `ref`. I think `withNavigation` would only work inside a child component of the `AppContainer`, since the `navigation` is only defined there.

Comment (maxammann) [reply src/modules/app/components/Navigator.js]: Yes that's what I meant. Not sure whether that works in this case though. The idea would be, as you said, to create a child component.

Afaik refs shouldn't be needed in 99% of the cases.

File: src/modules/app/StoreActionType.js [hunk 11]
```diff
@@ -123,6 +123,10 @@ export type ClearCityContentActionType = {|
   type: 'CLEAR_CITY_CONTENT'
 |}
 
+export type ClearCityActionType = {|
```
[the remark this answers is not in the document]
Comment (klinzo) [reply src/modules/app/StoreActionType.js]: We need two actions because we need to call asynchronous methods (`clearCitySelection`) and modify the state. So one is a saga to call the async method and one a normal redux reducer.

File: src/modules/app/components/Navigator.js [hunk 12]
```diff
@@ -74,12 +79,86 @@ export const AppStack = createStackNavigator(
...
+    this.navigator.current.dispatch(NavigationActions.navigate({
+      routeName: 'App',
+      action: navigateToDashboard
+    }))
+
+    this.props.fetchCategory(cityCode, language, key)
+  }
+
+  render () {
+    return <AppContainer ref={this.navigator} />
```
[the remark this answers is not in the document]
Comment (klinzo) [reply src/modules/app/components/Navigator.js]: Nope, just a white screen shows up. Should we create a task to show the spalsh-screen here?
